File: apps/ios/ProjectHmrClient.swift
import Foundation

enum ProjectHmrClient {
    private static let queue = DispatchQueue(label: "com.rayact.hmr", qos: .userInitiated)
    private static var generation = 0
    private static var running = false
    private static var webSocketTask: URLSessionWebSocketTask?
    private static var receiveTask: Task<Void, Never>?
    private static var baseUrl: String?
    private static weak var session: RayactEngineSession?

    static func start(serverUrl: String, engineSession: RayactEngineSession) {
        stop()
        let normalized = DevServerLoader.normalizeBase(serverUrl)
        generation += 1
        let gen = generation
        baseUrl = normalized
        session = engineSession
        running = true

        queue.async {
            guard running, gen == generation else { return }
            let manifest: [String: Any]
            do {
                manifest = try DevServerLoader.fetchManifest(baseUrl: normalized)
            } catch {
                print("[ProjectHmrClient] hmr manifest fetch failed: \(error.localizedDescription)")
                return
            }
            guard running, gen == generation else { return }

            let wsUrlString: String
            if let hmrUrl = manifest["hmrUrl"] as? String, !hmrUrl.isEmpty {
                wsUrlString = hmrUrl
            } else {
                wsUrlString = normalized
                    .replacingOccurrences(of: "http://", with: "ws://")
                    .replacingOccurrences(of: "https://", with: "wss://") + "/rayact/hmr"
            }
            guard let wsUrl = URL(string: wsUrlString) else { return }

            print("[ProjectHmrClient] connecting hmr ws=\(wsUrlString) gen=\(gen)")
            let task = URLSession.shared.webSocketTask(with: wsUrl)
            webSocketTask = task
            task.resume()
            listen(task: task, gen: gen, base: normalized, engineSession: engineSession)
        }
    }

    static func stop() {
        running = false
        generation += 1
        receiveTask?.cancel()
        receiveTask = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        session = nil
        baseUrl = nil
    }

    private static func listen(task: URLSessionWebSocketTask, gen: Int, base: String, engineSession: RayactEngineSession) {
        receiveTask = Task {
            while running, gen == generation, !Task.isCancelled {
                do {
                    let message = try await task.receive()
                    guard running, gen == generation else { return }
                    switch message {
                    case .string(let text):
                        handleMessage(base: base, engineSession: engineSession, raw: text)
                    case .data(let data):
                        if let text = String(data: data, encoding: .utf8) {
                            handleMessage(base: base, engineSession: engineSession, raw: text)
                        }
                    @unknown default:
                        break
                    }
                } catch {
                    print("[ProjectHmrClient] hmr ws failure: \(error.localizedDescription)")
                    return
                }
            }
        }
    }

    private static func handleMessage(base: String, engineSession: RayactEngineSession, raw: String) {
        guard let data = raw.data(using: .utf8),
              let message = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = message["type"] as? String else { return }

        switch type {
        case "update":
            guard let updates = message["updates"] as? [[String: Any]] else { return }
            for update in updates {
                guard (update["type"] as? String) == "js-update",
                      let path = update["path"] as? String, !path.isEmpty else { continue }
                let timestamp = update["timestamp"] as? Int64 ?? 0
                applyModuleUpdate(base: base, engineSession: engineSession, path: path, timestamp: timestamp)
            }
        case "full-reload", "reload":
            print("[ProjectHmrClient] full-reload requested")
            DevClientBridge.reloadCurrentProject()
        default:
            break
        }
    }

    private static func applyModuleUpdate(base: String, engineSession: RayactEngineSession, path: String, timestamp: Int64) {
        let canonicalPath = path.split(whereSeparator: { $0 == "?" || $0 == "#" }).first.map(String.init) ?? ""
        guard !canonicalPath.isEmpty else { return }
        let query = timestamp > 0 ? "?t=\(timestamp)&platform=ios" : "?platform=ios"
        let moduleUrl = "\(base)/rayact/m\(canonicalPath)\(query)"
        do {
            let source = try DevServerLoader.httpGetText(moduleUrl)
            let ok = engineSession.applyModuleUpdate(path: canonicalPath, source: source)
            print("[ProjectHmrClient] module update path=\(canonicalPath) ok=\(ok) bytes=\(source.count)")
        } catch {
            print("[ProjectHmrClient] module update failed path=\(canonicalPath): \(error)")
        }
    }
}

File: apps/android/app/src/main/java/com/rayact/devclient/ProjectHmrClient.kt
package com.rayact.devclient

import android.util.Log
import com.rayact.engine.RayactEngineSession
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

object ProjectHmrClient {
    private const val TAG = "ProjectHmrClient"
    private val client = OkHttpClient.Builder().build()
    private val generation = AtomicInteger(0)
    private val running = AtomicBoolean(false)

    @Volatile private var webSocket: WebSocket? = null
    @Volatile private var baseUrl: String? = null
    @Volatile private var session: RayactEngineSession? = null

    fun start(serverUrl: String, engineSession: RayactEngineSession) {
        stop()
        val normalized = DevServerLoader.normalizeBase(serverUrl)
        val gen = generation.incrementAndGet()
        baseUrl = normalized
        session = engineSession
        running.set(true)

        // Runs from runOnUiThread (openProject), so the manifest fetch must not
        // block the main thread — a synchronous httpGet here throws
        // NetworkOnMainThreadException, the connect is skipped, and HMR silently
        // never starts. Do the fetch + ws connect on a background thread.
        Thread {
            if (!running.get() || gen != generation.get()) return@Thread
            val manifest = runCatching {
                JSONObject(DevServerLoader.httpGetText("$normalized/rayact/manifest.json?platform=android"))
            }.getOrElse {
                Log.w(TAG, "hmr manifest fetch failed: ${it.message}")
                return@Thread
            }
            if (!running.get() || gen != generation.get()) return@Thread

            val wsUrl = manifest.optString("hmrUrl").ifBlank {
                normalized.replace("http://", "ws://").replace("https://", "wss://") + "/rayact/hmr"
            }

            Log.i(TAG, "connecting hmr ws=$wsUrl gen=$gen")
            val request = Request.Builder().url(wsUrl).build()
            webSocket = client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    Log.i(TAG, "hmr ws connected gen=$gen")
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    if (!running.get() || gen != generation.get()) return
                    handleMessage(normalized, engineSession, text)
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    Log.w(TAG, "hmr ws failure: ${t.message}")
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    Log.i(TAG, "hmr ws closed code=$code reason=$reason")
                }
            })
        }.start()
    }

    fun stop() {
        running.set(false)
        generation.incrementAndGet()
        webSocket?.close(1000, "stop")
        webSocket = null
        session = null
        baseUrl = null
    }

    private fun handleMessage(base: String, engineSession: RayactEngineSession, raw: String) {
        val message = runCatching { JSONObject(raw) }.getOrNull() ?: return
        when (message.optString("type")) {
            "update" -> {
                val updates = message.optJSONArray("updates") ?: return
                for (i in 0 until updates.length()) {
                    val update = updates.optJSONObject(i) ?: continue
                    if (update.optString("type") != "js-update") continue
                    val path = update.optString("path")
                    if (path.isBlank()) continue
                    val timestamp = update.optLong("timestamp", 0L)
                    applyModuleUpdate(base, engineSession, path, timestamp)
                }
            }
            "full-reload", "reload" -> {
                Log.i(TAG, "full-reload requested")
                DevClientBridge.reloadCurrentProject()
            }
        }
    }

    private fun applyModuleUpdate(base: String, engineSession: RayactEngineSession, path: String, timestamp: Long) {
        val query = if (timestamp > 0L) "?t=$timestamp&platform=android" else "?platform=android"
        val moduleUrl = "$base/rayact/m$path$query"
        try {
            val source = DevServerLoader.httpGetText(moduleUrl)
            val ok = engineSession.applyModuleUpdate(path, source)
            Log.i(TAG, "module update path=$path ok=$ok bytes=${source.length}")
        } catch (e: Exception) {
            Log.e(TAG, "module update failed path=$path", e)
        }
    }
}
